src/java.base/share/classes/sun/security/ssl/X509KeyManagerImpl.java

```diff
@@ -180,6 +301,116 @@ public String[] getServerAliases(String keyType, Principal[] issuers) {
     // implementation private methods
     //
 
+    private String removeAliasIndex (String alias){
+        if (alias == null) {
+            return null;
+        }
+
+        int firstDot = alias.indexOf('.');
+        int secondDot = alias.indexOf('.', firstDot + 1);
+        if ((firstDot == -1) || (secondDot == firstDot)) {
+            // invalid alias
+            return null;
+        }
+        return alias.substring(firstDot + 1);
+    }
+
+    private boolean areCertChainsEqual(String builderAlias,
+            X509Certificate[] cachedCerts) {
+        int aDot = builderAlias.indexOf('.');
+        int builderIndex = Integer.parseInt(builderAlias.substring(0, aDot));
+        String entryAlias = builderAlias.substring(aDot + 1);
+        try {
+            Builder builder = builders.get(builderIndex);
+            KeyStore keyStore = builder.getKeyStore();
+            if (keyStore == null) {
+                return false;
+            }
+            Certificate[] keyStoreCerts = keyStore.getCertificateChain(entryAlias);
+            if (keyStoreCerts == null || keyStoreCerts.length == 0) {
+                return false;
+            }
+
+            // Convert KeyStore certificates to X509Certificates
+            X509Certificate[] newCerts = Arrays.copyOf(keyStoreCerts,
+                    keyStoreCerts.length, X509Certificate[].class);
+            return Arrays.equals(cachedCerts, newCerts);
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+    private boolean processCredentials(Builder builder, int builderIndex,
+            String alias) throws Exception {
+        KeyStore keyStore = builder.getKeyStore();
+        if (keyStore == null) {
+            return false;
+        }
+
+        Entry newEntry = keyStore.getEntry(alias,
+                builder.getProtectionParameter(alias));
+        if (!(newEntry instanceof PrivateKeyEntry)) {
+            return false;
+        }
+
+        PrivateKeyEntry privateKeyEntry = (PrivateKeyEntry) newEntry;
+        PrivateKey privateKey = privateKeyEntry.getPrivateKey();
+        Certificate[] certs = privateKeyEntry.getCertificateChain();
```

Comment:
The method `getCertificateChain(String)` contains the below codes (line 211~213),
```
PrivateKeyEntry entry = getEntry(alias);
return entry == null ? null :
        (X509Certificate[])entry.getCertificateChain();
```
This method assumes `PrivateKeyEntry::getCertificateChain` should always return `X509Certificate[]`.

Then, could line 358 also cast the certificate chain to `X509Certificate[]` directly?
If so, the following codes can be simplified.
It may not need to check the certification type, and could not to convert the `Certificate[]` to `X509Certificate[]`.

Reply:
Changed to cast the certificate chain.